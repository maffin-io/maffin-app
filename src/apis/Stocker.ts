import { Amplify, API } from 'aws-amplify';
import { DateTime } from 'luxon';

import awsExports from '../aws-exports';

Amplify.configure(awsExports);

const API_NAME = 'stocker';

export async function getPrices(tickers: string[]): Promise<{ [key: string]:LiveSummary; }> {
  const options = {
    queryStringParameters: {
      ids: [tickers].toString(),
    },
  };
  const resp = await API.get(API_NAME, '/api/prices', options);

  return resp;
}

export async function getPrice(
  ticker: string,
  when: DateTime,
): Promise<{ price: number, currency: string }> {
  const options = {
    queryStringParameters: {
      id: ticker,
      when: Math.floor(when.toSeconds()),
    },
  };
  const resp = await API.get(API_NAME, '/api/price', options);

  return {
    price: Number(resp.price.toFixed(4)),
    currency: resp.currency,
  };
}

export type LiveSummary = {
  price: number,
  changePct: number,
  changeAbs: number,
  currency: string,
};
