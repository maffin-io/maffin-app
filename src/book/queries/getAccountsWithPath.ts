import { FindOptionsRelations, Not } from 'typeorm';
import { Account, Book } from '../entities';

type AccountPath = {
  guid: string,
  type: string,
  path: string,
  commodity: string,
  commodityGuid: string,
};

type GetAccountsWithPathProps = {
  showRoot?: boolean,
  relations?: FindOptionsRelations<Account>,
};

export async function getAccountsWithPath(
  {
    showRoot = false,
    relations = {},
  }: GetAccountsWithPathProps = {},
): Promise<Account[]> {
  const start = performance.now();
  const books = await Book.find();
  const rootAccount = books[0].root;

  const typeFilter = !showRoot ? { type: Not('ROOT') } : {};

  const [accountsPaths, accounts] = await Promise.all([
    Account.query(
      `
      WITH RECURSIVE absolute_names(guid, name, parent_guid, path) AS (
        SELECT guid, name, parent_guid, name FROM accounts WHERE parent_guid = $1
        UNION ALL
        SELECT a.guid, a.name, a.parent_guid, p.path || ':' || a.name
        FROM accounts AS a JOIN absolute_names AS p ON a.parent_guid = p.guid
      )
      SELECT guid, path 
      FROM absolute_names
      `,
      [rootAccount.guid],
    ),
    Account.find({
      where: typeFilter,
      relations,
    }),
  ]);

  accounts.forEach((account) => {
    const accountPath = accountsPaths.find(
      (ap: AccountPath) => ap.guid === account.guid,
    );

    if (showRoot && account.type === 'ROOT') {
      account.path = 'Root';
    } else {
      account.path = accountPath.path;
    }
  });

  const end = performance.now();
  console.log(`get accounts with paths: ${end - start}ms`);

  return accounts;
}
