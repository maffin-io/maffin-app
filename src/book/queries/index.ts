export { getInvestments } from './getInvestments';
export { getMainCurrency } from './getMainCurrency';
export { getEarliestDate } from './getEarliestDate';
export { getSplits } from './getSplits';
export { getLatestTxs } from './getLatestTxs';
