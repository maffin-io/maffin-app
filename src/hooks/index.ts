export { default as useDataSource, DataSourceContext } from './useDataSource';
export type { DataSourceContextType } from './useDataSource';
export { default as useBookStorage } from './useBookStorage';
export { default as useGapiClient } from './useGapiClient';
