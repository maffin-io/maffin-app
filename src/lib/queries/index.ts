export { default as getPrices } from './getPrices';

export { default as getMonthlyTotals } from './getMonthlyTotals';
export { default as getMainCurrency } from './getMainCurrency';
export { default as getEarliestDate } from './getEarliestDate';
export { default as getLatestTxs } from './getLatestTxs';
export { default as getAccountsTotals } from './getAccountsTotals';
