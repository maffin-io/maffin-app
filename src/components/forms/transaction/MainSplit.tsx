import React from 'react';
import { DateTime } from 'luxon';
import classNames from 'classnames';
import type { UseFormReturn } from 'react-hook-form';
import { Tooltip } from 'react-tooltip';
import { BiRightArrowAlt } from 'react-icons/bi';

import { isInvestment } from '@/book/helpers/accountType';
import { getPrice } from '@/apis/Stocker';
import type { Account, Commodity } from '@/book/entities';
import { toFixed } from '@/helpers/number';
import { currencyToSymbol } from '@/helpers/currency';
import type { FormValues } from './types';

export type MainSplitProps = {
  form: UseFormReturn<FormValues>,
  disabled?: boolean,
};

export default function MainSplit({
  form,
  disabled = false,
}: MainSplitProps): JSX.Element {
  const [exchangeRate, setExchangeRate] = React.useState(1);
  const account = form.getValues('splits.0.fk_account') as Account;
  const date = form.watch('date');
  const txCurrency = form.watch('fk_currency');
  const quantity = form.watch('splits.0.quantity');

  React.useEffect(() => {
    async function fetchRate() {
      const rate = await getExchangeRate(
        account,
        txCurrency,
        DateTime.fromISO(date),
      );
      setExchangeRate(rate);
    }

    if (form.formState.isDirty && date) {
      fetchRate();
    }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [txCurrency, date, form.formState.isDirty]);

  React.useEffect(() => {
    if (form.formState.isDirty) {
      form.setValue('splits.0.value', toFixed(quantity * exchangeRate, 3));
      form.trigger('splits');
    }
  }, [quantity, exchangeRate, form]);

  return (
    <div className="flex w-1/2">
      <div className="flex items-center rounded-md bg-light-100 dark:bg-dark-800">
        <input
          {...form.register(
            'splits.0.quantity',
            {
              valueAsNumber: true,
            },
          )}
          aria-label="splits.0.quantity"
          className="w-full text-right m-0"
          type="number"
          step="0.001"
          disabled={disabled}
        />
        <span className="pr-2">
          {currencyToSymbol(account?.commodity?.mnemonic || '').slice(0, 3)}
        </span>
      </div>

      <div
        className={
          classNames(
            'flex items-center rounded-md',
            {
              hidden: txCurrency.mnemonic === account.commodity.mnemonic,
            },
          )
        }
      >
        <span
          className="badge mx-3"
          data-tooltip-id="value-help"
        >
          <BiRightArrowAlt />
        </span>
        <Tooltip
          id="value-help"
          className="tooltip"
          disableStyleInjection
        >
          <p className="mb-2">
            The account is not in your main currency so this is the converted
            total amount for this transaction.
          </p>
        </Tooltip>

        <div
          className={
            classNames(
              'flex items-center rounded-md  bg-light-100 dark:bg-dark-800',
              {
                'outline outline-red-500/50': form.formState.errors.splits?.type === 'splitsBalance',
              },
            )
          }
        >
          <input
            {...form.register(
              'splits.0.value',
              {
                valueAsNumber: true,
              },
            )}
            aria-label="splits.0.value"
            className="w-full text-right m-0"
            type="number"
            step="0.001"
            disabled={disabled}
            hidden={txCurrency.mnemonic === account.commodity.mnemonic}
            onChange={(e) => {
              form.setValue('splits.0.value', Number(e.target.value));
              form.trigger('splits');
            }}
          />
          <span className="pr-2">
            {currencyToSymbol(txCurrency.mnemonic || '').slice(0, 3)}
          </span>
        </div>
      </div>
    </div>
  );
}

/**
 * Retrieves the rate to convert the current account commodity
 * to the currency of the transaction.
 *
 * If the account is an investment, we fetch its value at the given
 * date instead.
 */
async function getExchangeRate(
  account: Account,
  to: Commodity,
  when: DateTime,
): Promise<number> {
  if (account.commodity.mnemonic === to.mnemonic) {
    return 1;
  }
  let ticker = `${account.commodity.mnemonic}${to.mnemonic}=X`;

  if (isInvestment(account)) {
    ticker = account.commodity.mnemonic;
  }

  const rate = await getPrice(ticker, when);
  // TODO: We should check somewhere that the txCurrency is the same
  // as the stocks' currency to avoid price discrepancies.
  return rate.price;
}
