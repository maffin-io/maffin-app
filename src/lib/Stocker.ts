import { Amplify, API } from 'aws-amplify';
import { DateTime } from 'luxon';

import { Commodity, Price } from '@/book/entities';
import { getMainCurrency } from '@/lib/queries';
import { toAmountWithScale } from '@/helpers/number';

import awsExports from '../aws-exports';

Amplify.configure(awsExports);

const API_NAME = 'stocker';

/**
 * Connect to Stocker API and retrieve current prices for
 * all currency pairs and commodities.
 */
export async function insertTodayPrices(): Promise<void> {
  const start = performance.now();
  const mainCurrency = await getMainCurrency();
  const [currencies, commodities] = await Promise.all([
    Commodity.findBy({ namespace: 'CURRENCY' }),
    Commodity.find({
      where: [
        { namespace: 'STOCK' },
        { namespace: 'MUTUAL' },
      ],
    }),
  ]);

  const currencyTickers = currencies
    .filter(c => c.guid !== mainCurrency.guid)
    .map(c => `${c.mnemonic}${mainCurrency.mnemonic}=X`);

  const commodityTickers = commodities.map(c => c.cusip || c.mnemonic);
  const tickers = [
    ...currencyTickers,
    ...commodityTickers,
  ];

  const options = {
    queryStringParameters: {
      ids: Array.from(new Set(tickers)).toString(),
    },
  };
  const resp = await API.get(API_NAME, '/api/prices', options) as { [key: string]:LiveSummary; };

  const now = DateTime.now().startOf('day');

  const prices = Object.entries(resp).map(([key, summary]) => {
    let commodityMnemonic = key.substring(0, 3);
    let currencyMnemonic = key.substring(3, 6);
    if (!key.endsWith('=X')) {
      commodityMnemonic = key;
      currencyMnemonic = summary.currency;
    }
    const { amount, scale } = toAmountWithScale(summary.price);

    const commodity: Commodity = (
      currencies.find(c => c.mnemonic === commodityMnemonic)
      || commodities.find(
        c => (c.mnemonic === commodityMnemonic || c.cusip === commodityMnemonic),
      )
    ) as Commodity;

    const currency: Commodity = currencies.find(c => c.mnemonic === currencyMnemonic) as Commodity;

    return Price.create({
      fk_commodity: commodity.guid,
      fk_currency: currency.guid,
      date: now,
      source: `maffin::${JSON.stringify(summary)}`,
      valueNum: amount,
      valueDenom: parseInt('1'.padEnd(scale + 1, '0'), 10),
    });
  });

  await Price.upsert(
    prices,
    {
      conflictPaths: ['fk_commodity', 'fk_currency', 'date'],
    },
  );

  if (prices.length) {
    prices[0]?.queryClient?.invalidateQueries({
      queryKey: ['api', 'prices'],
    });
  }

  const end = performance.now();
  console.log(`/stocker/api/prices: ${end - start}ms`);
}

export type LiveSummary = {
  price: number,
  changePct: number,
  changeAbs: number,
  currency: string,
};
