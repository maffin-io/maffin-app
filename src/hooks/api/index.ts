import { DateTime } from 'luxon';
import { mutate, SWRResponse } from 'swr';
import useSWRImmutable from 'swr/immutable';

import { Commodity } from '@/book/entities';
import { InvestmentAccount } from '@/book/models';
import * as queries from '@/lib/queries';
import type { PriceDBMap } from '@/book/prices';
import type { AccountsMap } from '@/types/book';
import type { Account, Split, Transaction } from '@/book/entities';

export function useStartDate(): SWRResponse<DateTime> {
  const key = '/api/start-date';
  return useSWRImmutable(
    key,
    fetcher(queries.getEarliestDate, key),
  );
}

export function useMainCurrency(): SWRResponse<Commodity> {
  const key = '/api/main-currency';
  return useSWRImmutable(
    key,
    fetcher(queries.getMainCurrency, key),
  );
}

export function useCommodity(guid: string): SWRResponse<Commodity> {
  const key = `/api/commodities/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => Commodity.findOneByOrFail({ guid }), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useCommodities(): SWRResponse<Commodity[]> {
  const key = '/api/commodities';

  const result = useSWRImmutable(
    key,
    fetcher(() => Commodity.find(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    result.data.forEach(
      (commodity: Commodity) => mutate(
        `/api/commodities/${commodity.guid}`,
        commodity,
        { revalidate: false },
      ),
    );
  }

  return result;
}

export function useLatestTxs(): SWRResponse<Transaction[]> {
  const key = '/api/txs/latest';
  return useSWRImmutable(
    key,
    fetcher(queries.getLatestTxs, key),
  );
}

export function useAccount(guid: string): SWRResponse<Account> {
  const key = `/api/accounts/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getAccounts(guid), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useAccounts(): SWRResponse<AccountsMap> {
  const key = '/api/accounts';

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getAccounts(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    Object.values(result.data as AccountsMap).forEach(
      (account: Account) => mutate(`/api/accounts/${account.guid}`, account, { revalidate: false }),
    );
  }

  return result;
}

export function useAccountsMonthlyTotals(): SWRResponse<queries.MonthlyTotals> {
  const { data: accounts } = useAccounts();
  const { data: prices } = useSWRImmutable(
    '/api/prices',
    fetcher(() => queries.getPrices({}), '/api/prices'),
  );

  const key = '/api/monthly-totals';
  const result = useSWRImmutable(
    (accounts && prices) ? key : null,
    fetcher(
      () => queries.getMonthlyTotals(accounts as AccountsMap, prices),
      key,
    ),
  );

  return result;
}

export function useInvestment(guid: string): SWRResponse<InvestmentAccount> {
  const key = `/api/investments/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getInvestment(guid), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useInvestments(): SWRResponse<InvestmentAccount[]> {
  const key = '/api/investments';

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getInvestments(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    result.data.forEach(
      (investment: InvestmentAccount) => mutate(
        `/api/investments/${investment.account.guid}`,
        investment,
        { revalidate: false },
      ),
    );
  }

  return result;
}

export function useSplits(guid: string): SWRResponse<Split[]> {
  const key = `/api/splits/${guid}`;
  return useSWRImmutable(
    key,
    fetcher(async () => queries.getSplits(guid), key),
  );
}

/**
 * Returns prices for a given commodity
 */
export function usePrices(c: Commodity | undefined): SWRResponse<PriceDBMap> {
  const key = `/api/prices/${c?.guid}`;
  return useSWRImmutable(
    c?.guid ? key : null,
    fetcher(async () => queries.getPrices({ from: c }), key),
  );
}

function fetcher(f: () => Promise<any>, key: string) {
  return async () => {
    const start = performance.now();
    const r = await f();
    const end = performance.now();
    console.log(`${key}: ${end - start}ms`);
    return r;
  };
}
