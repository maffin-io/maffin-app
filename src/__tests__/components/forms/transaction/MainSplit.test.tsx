import React from 'react';
import {
  waitFor,
  render,
  screen,
} from '@testing-library/react';
import { DateTime } from 'luxon';
import userEvent from '@testing-library/user-event';
import { useForm } from 'react-hook-form';
import type { SWRResponse } from 'swr';

import Stocker from '@/apis/Stocker';
import MainSplit from '@/components/forms/transaction/MainSplit';
import * as queries from '@/lib/queries';
import * as apiHook from '@/hooks/api';
import type { Account, Commodity } from '@/book/entities';
import type { FormValues } from '@/components/forms/transaction/types';

jest.mock('@/lib/queries', () => ({
  __esModule: true,
  ...jest.requireActual('@/lib/queries'),
}));

jest.mock('@/hooks/api', () => ({
  __esModule: true,
  ...jest.requireActual('@/hooks/api'),
}));

describe('MainSplit', () => {
  let eur: Commodity;
  const EURSGD = 1.44;

  beforeEach(() => {
    eur = {
      guid: 'eur',
      mnemonic: 'EUR',
    } as Commodity;

    jest.spyOn(queries, 'getMainCurrency').mockResolvedValue(eur);
    jest.spyOn(apiHook, 'useAccounts')
      .mockReturnValue({ data: undefined } as SWRResponse);
    jest.spyOn(Stocker.prototype, 'getPrice').mockResolvedValue({ price: EURSGD, currency: 'EUR' });
  });

  it('renders with empty data', () => {
    const { container } = render(<FormWrapper />);
    expect(screen.getByRole('spinbutton', { name: 'splits.0.quantity' })).toBeVisible();
    expect(screen.getByRole('spinbutton', { name: '', hidden: true })).not.toBeVisible();
    expect(container).toMatchSnapshot();
  });

  it('renders disabled', () => {
    render(<FormWrapper disabled />);

    expect(screen.getByRole('spinbutton', { name: 'splits.0.quantity' })).toBeDisabled();
    expect(screen.getByRole('spinbutton', { name: '', hidden: true })).toBeDisabled();
  });

  it('shows value field when txCurrency is different', async () => {
    render(
      <FormWrapper
        defaults={{
          fk_currency: { mnemonic: 'SGD' } as Commodity,
        } as FormValues}
      />,
    );

    await waitFor(() => {
      expect(screen.getByRole('spinbutton', { name: 'splits.0.value' })).toBeVisible();
    });
  });

  // When we load data through defaults, we want to keep it as
  // we are loading the split for deleting or updating
  it('loads with default and doesnt reset value', async () => {
    const mockGetPrice = jest.spyOn(Stocker.prototype, 'getPrice')
      .mockResolvedValue({ price: EURSGD, currency: '' });

    render(
      <FormWrapper
        defaults={{
          fk_currency: { mnemonic: 'SGD' } as Commodity,
          splits: [
            {
              quantity: 50,
              value: 100,
              fk_account: {
                name: 'path1',
                guid: 'account_guid_1',
                path: 'path1',
                type: 'STOCK',
                commodity: {
                  guid: 'eur',
                  mnemonic: 'EUR',
                } as Commodity,
              } as Account,
            },
          ],
        } as FormValues}
      />,
    );

    expect(screen.getByRole('spinbutton', { name: 'splits.0.value' })).toHaveValue(100);
    expect(screen.getByRole('spinbutton', { name: 'splits.0.quantity' })).toHaveValue(50);
    expect(mockGetPrice).not.toBeCalled();
  });

  it('sets value when quantity changes', async () => {
    const user = userEvent.setup();
    render(<FormWrapper />);

    const q0 = screen.getByRole('spinbutton', { name: 'splits.0.quantity' });
    const v0 = screen.getByRole('spinbutton', { name: '', hidden: true });

    expect(q0).toBeEnabled();
    expect(v0).toBeEnabled();
    await user.type(q0, '100');

    await waitFor(() => expect(q0).toHaveValue(100));
    await waitFor(() => expect(v0).toHaveValue(100));
  });

  it('sets value when quantity changes with different currency', async () => {
    const user = userEvent.setup();
    const mockGetPrice = jest.spyOn(Stocker.prototype, 'getPrice')
      .mockResolvedValue({ price: EURSGD, currency: 'EUR' });

    render(
      <FormWrapper
        defaults={{
          fk_currency: { mnemonic: 'SGD' } as Commodity,
        } as FormValues}
      />,
    );

    const q0 = screen.getByRole('spinbutton', { name: 'splits.0.quantity' });
    const v0 = screen.getByRole('spinbutton', { name: 'splits.0.value' });

    expect(q0).toBeEnabled();
    expect(v0).toBeEnabled();
    await user.type(q0, '100');

    await waitFor(() => expect(q0).toHaveValue(100));
    await waitFor(() => expect(v0).toHaveValue(100 * EURSGD));
    expect(mockGetPrice).toHaveBeenLastCalledWith('EURSGD=X', DateTime.fromISO('2023-01-01'));
  });

  it('retrieves exchange rate for investment', async () => {
    const user = userEvent.setup();
    const mockGetPrice = jest.spyOn(Stocker.prototype, 'getPrice')
      .mockResolvedValue({ price: 100.1, currency: 'USD' });

    render(
      <FormWrapper
        defaults={{
          fk_currency: { mnemonic: 'SGD' } as Commodity,
          splits: [
            {
              fk_account: {
                name: 'path1',
                guid: 'account_guid_1',
                path: 'path1',
                type: 'STOCK',
                commodity: {
                  guid: 'googl',
                  mnemonic: 'GOOGL',
                } as Commodity,
              } as Account,
            },
          ],
        } as FormValues}
      />,
    );

    const q0 = screen.getByRole('spinbutton', { name: 'splits.0.quantity' });
    const v0 = screen.getByRole('spinbutton', { name: 'splits.0.value' });

    expect(q0).toBeEnabled();
    expect(v0).toBeEnabled();
    await user.type(q0, '100');

    await waitFor(() => expect(q0).toHaveValue(100));
    await waitFor(() => expect(v0).toHaveValue(100 * 100.1));
    expect(mockGetPrice).toHaveBeenLastCalledWith('GOOGL', DateTime.fromISO('2023-01-01'));
  });

  it('recalculates when date changes', async () => {
    const user = userEvent.setup();
    const mockGetPrice = jest.spyOn(Stocker.prototype, 'getPrice')
      .mockResolvedValueOnce({ price: 1.30, currency: '' });

    render(
      <FormWrapper
        defaults={{
          fk_currency: { mnemonic: 'SGD' } as Commodity,
        } as FormValues}
      />,
    );

    const dateField = screen.getByTestId('date');
    const q0 = screen.getByRole('spinbutton', { name: 'splits.0.quantity' });
    const v0 = screen.getByRole('spinbutton', { name: 'splits.0.value' });

    user.clear(dateField);
    await user.type(q0, '100');
    await waitFor(() => expect(q0).toHaveValue(100));

    await user.type(dateField, '2023-01-20');
    await waitFor(() => expect(v0).toHaveValue(130));
    await waitFor(
      () => expect(mockGetPrice).toHaveBeenLastCalledWith('EURSGD=X', DateTime.fromISO('2023-01-20')),
    );
  });
});

function FormWrapper(
  {
    disabled = false,
    defaults = {} as FormValues,
  }: {
    disabled?: boolean,
    defaults?: FormValues,
  },
): JSX.Element {
  const defaultValues = {
    date: '2023-01-01',
    splits: [
      {
        value: 0,
        quantity: 0,
        fk_account: {
          name: 'path1',
          guid: 'account_guid_1',
          path: 'path1',
          type: 'ASSET',
          commodity: {
            guid: 'eur',
            mnemonic: 'EUR',
          } as Commodity,
        } as Account,
      },
    ],
    fk_currency: { guid: 'eur', mnemonic: 'EUR' },
  };

  const form = useForm<FormValues>({
    // These values reproduce the same initial state TransactionForm renders
    defaultValues: {
      ...defaultValues,
      ...defaults,
    },
  });

  // calling trigger creates some issues with background rendering for tests
  form.trigger = jest.fn();

  return (
    <>
      <input
        id="dateInput"
        data-testid="date"
        className="block w-full m-0"
        {...form.register('date')}
        type="date"
      />
      <MainSplit form={form} disabled={disabled} />
    </>
  );
}
