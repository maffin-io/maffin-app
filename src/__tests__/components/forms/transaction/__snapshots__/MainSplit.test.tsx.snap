// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainSplit renders with empty data 1`] = `
<div>
  <input
    class="block w-full m-0"
    data-testid="date"
    id="dateInput"
    name="date"
    type="date"
  />
  <div
    class="flex w-1/2"
  >
    <div
      class="flex items-center rounded-md bg-background-800"
    >
      <input
        aria-label="splits.0.quantity"
        class="w-full text-right m-0"
        name="splits.0.quantity"
        step="any"
        type="number"
      />
      <span
        class="pr-2"
      >
        €
      </span>
    </div>
    <div
      class="flex items-center rounded-md hidden"
    >
      <span
        class="badge default mx-3"
        data-tooltip-id="value-help"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m11.293 17.293 1.414 1.414L19.414 12l-6.707-6.707-1.414 1.414L15.586 11H6v2h9.586z"
          />
        </svg>
      </span>
      <div
        class="flex items-center rounded-md  bg-background-800"
      >
        <input
          aria-label="splits.0.value"
          class="w-full text-right m-0"
          hidden=""
          name="splits.0.value"
          step="0.001"
          type="number"
        />
        <span
          class="pr-2"
        >
          €
        </span>
      </div>
    </div>
  </div>
</div>
`;
